db: fix SQL TRANSACTION syntax compatible with all DB backends

tmszi opened the pull request:
SQL TRANSACTION syntax (compatible with all DB backends):

```
BEGIN;
...;
COMMIT;
```

This default SQL TRANSACTION syntax is not compatible with [MySQL/MariaDB DB backend](https://dev.mysql.com/doc/refman/8.0/en/commit.html):

```
BEGIN TRANSACTION;
...;
END TRANSACTION;
```

Successfully tested with SQLite, PostgreSQL, MySQL DB backend.

tmszi committed:
db: fix SQL TRANSACTION syntax compatible with all DB backends

SQL TRANSACTION syntax:

```
BEGIN;
...;
COMMIT;
```

nilason commented:
I wonder if this shouldn't be approached in a different way. There _is_ C API (driver specific) for this (`db_begin_transaction()`, `db_commit_transaction()`).

Anyways, it should have been `START TRANSACTION` and `COMMIT`, not `BEGIN TRANSACTION` and `END TRANSACTION`.

tmszi commented:
> Anyways, it should have been `START TRANSACTION` and `COMMIT`, not `BEGIN TRANSACTION` and `END TRANSACTION`.

or BEGIN alias for START TRANSACTION.

nilason commented:
> > Anyways, it should have been `START TRANSACTION` and `COMMIT`, not `BEGIN TRANSACTION` and `END TRANSACTION`.
> 
> or BEGIN as alias for START TRANSACTION.

Yes, for the drivers that support it. You may take a look at how this is implemented for the different drivers in `db/drivers/*/execute.c` files.

marcjansen commented on scripts/db.dropcolumn/db.dropcolumn.py:
```diff
@@ -120,14 +120,14 @@ def main():
 
             cmds = [
-                "BEGIN TRANSACTION",
+                "BEGIN;",
```
The added `;` will result in duplicated `;` since in line 132 the semicolons are added for all elements in the list. Probably just cosmetic, but might be better to not add the `;` here. 

This also applies to line 130, and also `v.db.dropcolumn.py`

tmszi replied:
Thanks!

tmszi commented:
> I wonder if this shouldn't be approached in a different way. There _is_ C API (driver specific) for this (`db_begin_transaction()`, `db_commit_transaction()`).

Yes you are right we can refactor it to better implementation on the low C lang level.

tmszi committed:
db.execute: start transaction before execute SQL and commit after

tmszi committed:
Remove SQL BEGIN and COMMIT command transaction

nilason commented:
Injecting sql code in `db.execute` breaks current behaviour, so that is not a good option. I was originally thinking of directly use C API in Python, like in:
https://github.com/OSGeo/grass/blob/7413740dd81c11c2909ad10e7c3161fc097088f9/gui/wxpython/vdigit/wxdigit.py#L575
or
https://github.com/OSGeo/grass/blob/7413740dd81c11c2909ad10e7c3161fc097088f9/gui/wxpython/vdigit/wxdigit.py#L1246

but that will require quite a lot of rewriting.

But you could perhaps use:
https://github.com/OSGeo/grass/blob/7413740dd81c11c2909ad10e7c3161fc097088f9/python/grass/script/db.py

https://github.com/OSGeo/grass/blob/7413740dd81c11c2909ad10e7c3161fc097088f9/python/grass/script/db.py#L235

tmszi committed:
Revert "db.execute: start transaction before execute SQL and commit after"

This reverts commit 06bbc2353ebfae69f8fec9bb5e620b7d685d99be.

tmszi committed:
python/grass/script: refactor py db_begin_transaction(), db_commit_transaction() func

To use C API db_begin_transaction(), db_commit_transaction().

Used by modules:

- db.dropcolumn
- v.db.addcolumn
- v.db.dropcolumn
- v.db.join.py
- v.db.univar
- v.dissolve
- v.rast.stats

tmszi committed:
Concatenate trans string

tmszi committed:
Exchange dbDriver* pointer between db_begin_transaction() and db_commit_transaction() and release opened driver/db connection

tmszi committed:
Add/use db_execute() func for reusing opened DB driver/connection

tmszi committed:
Fix removed tmp var

tmszi committed:
Fix make/print/append SQL string

tmszi committed:
Fix substituting SQLite DB path // var

tmszi committed:
Check if conn var is not None

tmszi committed:
Fix f-string syntax

nilason commented:
I like where this is going. Something like this is what I had in mind, although it is way beyond the original intent of this PR. But I don't mind. Allows for more flexibility and control using C API directly, than via the module `db.execute`.
Not sure what to do about the change of behaviour of `db_commit_transaction`(API break).
I can also imagine making this a class, reducing the need to pass driver, database etc to every function call, and possibility to push sql commands (instead of building a string)... but before you put more work into this let us ask for the general opinion of this change by e.g. @petrasovaa,  @wenzeslaus, @marisn.

tmszi committed:
Fix append() method name

tmszi commented on scripts/v.db.join/v.db.join.py:
```diff
@@ -210,31 +198,28 @@ def main():
-        update_str += cur_up_str
-    update_str += "END TRANSACTION"
-    grass.debug(update_str, 1)
+        sqls.apend(cur_up_str)
```
thanks, fixed with https://github.com/OSGeo/grass/pull/3636/commits/a8fec414b71c72604303dd7d850425c56500bbdd.

wenzeslaus commented:
I also would like to replace putting the strings together by something better given our C database API and Python database APIs. While alternatives exists and some are employed in the code already, using the C library directly through C types is a straightforward option.

The API change is a potential issue. How big, I don't know, I'm guessing not big. Still, breaking it seems wrong. You also mentioned the issue of passing multiple parameters. I would certainly try how this would look as a class. This would be one way to avoid the change in the API. Perhaps following [PEP 249 – Python Database API Specification](https://peps.python.org/pep-0249/) or mimicking relevant parts of an existing SQL database Python API (Psycopg, sqlite3, SQLAlchemy, ...) would be good. Ideally, such API would actually hide whatever is happening in the background (constructing string for db.execute, passing to a native wrapper like sqlite3 or passing directly, or through a subprocess, to C API).

While when ctypes works, everything is just perfect, the experience is that something is often missing from the things which need to be in place for C API to work in Python. Hence, we have all the lazy imports in this PR and elsewhere. That's a thing which can be gradually improved, though. The many updates to ctypes lately were certainly helpful. Switching to Filesystem Hierarchy Standard might be another update which would help here ([[GRASS-dev] GRASS 8.0 support in GDAL and QGIS](https://lists.osgeo.org/pipermail/grass-dev/2022-February/095596.html)).

I'm curious what others think.

marisn commented on scripts/v.db.dropcolumn/v.db.dropcolumn.py:
```diff
@@ -122,9 +122,15 @@ def main():
 
         try:
+            pdriver = db_begin_transaction(driver_name=driver, database=database)
             grass.write_command(
```
Why not `db_execute` as in other cases?

marisn commented:
> I also would like to replace putting the strings together by something better given our C database API and Python database APIs. While alternatives exists and some are employed in the code already, using the C library directly through C types is a straightforward option.

Proper prepared statements are way to go but lets take one step at a time (that would require a new C and Python API).

> The API change is a potential issue. How big, I don't know, I'm guessing not big. Still, breaking it seems wrong. 

This wouldn't be the first time when we break API on a minor release. Taking into account that this is not the most commonly used functionality, it fixes compatibility with one of our supported back-ends, I would not cry over this.

echoix commented:
Is this still planned for the due 8.4.0 release?

tmszi committed:
Refactor code to use DBHandler class instance

nilason commented:
https://github.com/OSGeo/grass/pull/3636/commits/093c461380326dccd897db142fc603d50aea2490: great initiative!

tmszi committed:
Fix Flake8 E501 line too long (89 > 88 characters) error

echoix commented on python/grass/script/db.py:
```diff
@@ -233,23 +246,139 @@ def db_table_in_vector(table, mapset=".", env=None):
+            self._c_funcs["db_close_database_shutdown_driver"](self._pdriver)
+            fatal(
+                _(
+                    "Error while start database <{db}> transaction by"
+                    " driver <{driver}>."
```
The `start` sounds weird. Is `starting` the correct wording?
```suggestion
                    "Error while starting database <{db}> transaction by"
                    " driver <{driver}>."
```

wenzeslaus commented on python/grass/script/db.py:
```diff
@@ -233,23 +246,139 @@ def db_table_in_vector(table, mapset=".", env=None):
+                                   statement
+        """
+        # Begin DB transaction
+        self._begin_transaction()
+        # Execute SQL string
+        if isinstance(sql, (list, tuple)):
+            for statement in sql:
+                self._execute(sql=statement)
+        else:
+            self._execute(sql)
+        # Commit DB transaction
+        self._commit_transaction()
```
While consistent within this code, database driver start and shutdown within one function are perhaps little unexpected. One probably should not do more than couple execute calls per process otherwise it might be a performance hit which is a big reason to use transactions in our case. This needs clear documentation.

wenzeslaus commented on scripts/v.db.addcolumn/v.db.addcolumn.py:
```diff
@@ -40,30 +40,15 @@
 def main():
+    from grass.script.db import DBHandler
+
     global rm_files
```
Delete.

BTW, I'm surprised this surpassed automatic code quality checks.

wenzeslaus commented on scripts/v.db.dropcolumn/v.db.dropcolumn.py:
```diff
@@ -37,12 +37,13 @@
 
 
 def main():
+    from grass.script.db import DBHandler
```
Any reason for lazy import? The current DBHandler does the tricky stuff in instantiation, not import time.

wenzeslaus reviewed (changes requested):
While I have doubts about reliability of ctypes interface (I guess too many bad experiences in the past), I appreciate the whole effort and the class makes sense.

wenzeslaus commented:
While very valuable, we can't fit this into 8.4, moving milestone to 8.5. I hope that's okay.

tmszi committed:
Incorporate suggestions

tmszi commented:
All suggestions was incorporated  into commit no https://github.com/OSGeo/grass/pull/3636/commits/072d0be7622dc177d6e5c993017577060be80923.

echoix commented:
There's now some conflicts to solve here

echoix committed:
Merge branch 'main' into fix-sql-transaction-syntax-compatible-with-all-db-backends

echoix committed:
style: Remove unused pathlib import

echoix committed:
typing: Change type in docstring to typing annotations

echoix commented:
Solved conflicts. The thing to look out for was that some commands are now using env=env, but the new class used here looses that info, it doesn't use it.
